pkgs/development/tools/analysis/ida-free/default.nix

```diff
@@ -0,0 +1,114 @@
+{ lib
+, stdenv
+, fetchurl
+, libGL
+, zlib
+, libsecret
+, libxkbcommon
+, glib
+, makeWrapper
+, openssl
+, xorg
+, dbus
+, fontconfig
+, freetype
+, makeDesktopItem
+}:
+stdenv.mkDerivation rec {
+  pname = "ida-free";
+  version = "8.2.230124";
+
+  src = fetchurl {
+    url = "https://web.archive.org/web/20230213091148if_/https://out7.hex-rays.com/files/idafree82_linux.run";
+    sha256 = "sha256-pcyb8M39FJVsWUAQiXy17nOtVUhPkPsE24YqhK9rK58=";
+  };
```

Comment:
update.sh

```sh
#! /usr/bin/env nix-shell
#! nix-shell -i bash -p bash coreutils curl jq

# faster than nix-shell:
#! /usr/bin/env bash

# https://hex-rays.com/ida-free/

# archive.org sends the original "last-modified" header as
# x-archive-orig-last-modified: Thu, 29 Dec 2022 14:45:57 GMT

urls=()
urls+=(https://out7.hex-rays.com/files/idafree82_windows.exe)
urls+=(https://out7.hex-rays.com/files/idafree82_linux.run)
urls+=(https://out7.hex-rays.com/files/idafree82_mac.app.zip)
urls+=(https://out7.hex-rays.com/files/arm_idafree82_mac.app.zip)

hashPrefix=sha256:

cd "$(dirname "$0")"

source_json_file=src.json

old_source_json='{}'
if [ -e $source_json_file ]; then
    # parse old source info
    old_source_json="$(< $source_json_file)"
fi

new_source_json="$old_source_json"

# parse hashes
if true; then
html=$(curl -s https://hex-rays.com/ida-free/)
else
html=$(cat <<'EOF'
test
https://hex-rays.com/ida-free/
b77a56913d669240b53f1cfd2deec20c3f4b00cbf286bedae46172e11f06431a  arm_idafree82_mac.app.zip
a5cc9bf0cdfd14956c594010897cb5ee73ad55484f90fb04db862a84af6b2b9f  idafree82_linux.run
60cd970c6aa01e14f488ef28c6017eca28bf447509f40176a1a982bb8e90f12a  idafree82_mac.app.zip
2289a03c9f2f2c2fa3de86453139302108a32abf3b98d27267b2c9bfa48056cf  idafree82_windows.exe
EOF
)
fi

hash_regex='[0-9a-f]{64}' # sha256base16

hashes_raw=$(echo "$html" | grep -E "^$hash_regex  [a-z0-9_.]*idafree[a-z0-9_.]+\.(run|zip|exe)$")

echo "hashes_raw:"
echo "$hashes_raw"

declare -A hashes
while read line; do
    sha256base16=$(echo "$line" | sed -E "s/^($hash_regex)  (.+)$/\1/")
    file=$(echo "$line" | sed -E "s/^($hash_regex)  (.+)$/\2/")
    hashes[$file]=$sha256base16
done <<< "$hashes_raw"

for url in "${urls[@]}"; do
    file=$(basename "$url")

    # parse platform from $file
    # arm_idafree82_mac.app.zip
    if echo "$file" | grep -qE '^arm_idafree[0-9]+_mac.app.zip$'; then
        platform=armv7a-darwin
    # idafree82_mac.app.zip
    elif echo "$file" | grep -qE '^idafree[0-9]+_mac.app.zip$'; then
        platform=x86_64-darwin
    # idafree82_linux.run
    elif echo "$file" | grep -qE '^idafree[0-9]+_linux.run$'; then
        platform=x86_64-linux
    # idafree82_windows.exe
    elif echo "$file" | grep -qE '^idafree[0-9]+_windows.exe$'; then
        platform=x86_64-windows
    else
        platform=fixme-unknown-platform-of-file-$file
    fi

    sha256base16=${hashes[$file]}
    hash="$hashPrefix$sha256base16"
    hashOld="$(echo "$old_source_json" | jq -r '.[$platform].hash' --arg platform "$platform")"
    #sha256base16Old=${hashOld/$hashPrefix/}

    if [[ "$hash" == "$hashOld" ]]
    then
        echo "source is up-to-date (same hash)"
        continue
    fi

    echo "hash changed:"
    echo "-hash: $hashOld"
    echo "+hash: $hash"

    contentLengthOld="$(echo "$old_source_json" | jq -r '.[$platform].contentLength' --arg platform "$platform")"
    lastModifiedOld="$(echo "$old_source_json" | jq -r '.[$platform].lastModified' --arg platform "$platform")"

    echo fetching headers
    headers=$(curl -s -I $url | tr -d $'\r')
    contentLength=$(echo "$headers" | grep "^Content-Length:" | tail -n1)
    contentLength=''${contentLength:16}
    lastModified=$(echo "$headers" | grep "^Last-Modified:" | tail -n1)
    lastModified=''${lastModified:15}

    versionDate=$(date --utc -d"$lastModified" +"%F")
    lastModified=$(date --utc -d"$lastModified" +"%FT%T%z") # "iso strict" format

    # parse version from $file
    versionNumber=$(echo "$file" | sed -E 's/^.*idafree([0-9])([0-9])_.*$/\1.\2/')

    version="$versionNumber-unstable-$versionDate"

    echo "$lastModified $url"

    # hash changed
    if false; then
    if [[ "$contentLengthOld" == "$contentLength" && "$lastModifiedOld" == "$lastModified" ]]
    then
        echo "source is up-to-date"
        continue
    fi
    fi

    #echo "original url changed:"
    if [[ "$contentLengthOld" != "$contentLength" ]]; then
        echo "-Content-Length: $contentLengthOld"
        echo "+Content-Length: $contentLength"
    fi
    if [[ "$lastModifiedOld" != "$lastModified" ]]; then
        echo "-Last-Modified: $lastModifiedOld"
        echo "+Last-Modified: $lastModified"
    fi

    # update source info
    new_source_json="$(
        echo "$new_source_json" | jq --sort-keys --slurp '
            [.[0], {
                ($platform): {
                    "url": $url,
                    "hash": $hash,
                    "version": $version,
                    "contentLength": $contentLength,
                    "lastModified": $lastModified
                }
            }] | add
        ' \
        --argjson contentLength "$contentLength" \
        --arg lastModified "$lastModified" \
        --arg hash "$hash" \
        --arg version "$version" \
        --arg platform "$platform" \
        --arg url "TODO: https://web.archive.org/save/$url and https://web.archive.org/web/*/$url" \
    )"
    #echo "new_source_json:"; echo "$new_source_json"

done

#echo "new_source_json:"; echo "$new_source_json"

echo updating $source_json_file
echo "$new_source_json" >$source_json_file
```

src.json

```json
{
  "armv7a-darwin": {
    "contentLength": 78392661,
    "hash": "sha256:08f10c8509056d04e21d79f04a5428441cd20dda275d8d20529bd08d7d7a9831",
    "lastModified": "2022-12-29T14:45:56+0000",
    "url": "https://web.archive.org/web/20221231191126/https://out7.hex-rays.com/files/arm_idafree82_mac.app.zip",
    "version": "8.2-unstable-2022-12-29"
  },
  "x86_64-darwin": {
    "contentLength": 76162919,
    "hash": "sha256:06c5dbb2093b09d27bcbe609dcd094b8c09a4b88bcb2909adb2b85b9ea68072d",
    "lastModified": "2022-12-29T14:45:57+0000",
    "url": "https://web.archive.org/web/20221231191107/https://out7.hex-rays.com/files/idafree82_mac.app.zip",
    "version": "8.2-unstable-2022-12-29"
  },
  "x86_64-linux": {
    "contentLength": 86388748,
    "hash": "sha256:a5cc9bf0cdfd14956c594010897cb5ee73ad55484f90fb04db862a84af6b2b9f",
    "lastModified": "2023-01-24T14:55:10+0000",
    "url": "https://web.archive.org/web/20230213091148/https://out7.hex-rays.com/files/idafree82_linux.run",
    "version": "8.2-unstable-2023-01-24"
  },
  "x86_64-windows": {
    "contentLength": 99596744,
    "hash": "sha256:2289a03c9f2f2c2fa3de86453139302108a32abf3b98d27267b2c9bfa48056cf",
    "lastModified": "2023-01-24T14:55:13+0000",
    "url": "https://web.archive.org/web/20230213074043/https://out7.hex-rays.com/files/idafree82_windows.exe",
    "version": "8.2-unstable-2023-01-24"
  }
}
```

default.nix

```nix
{ lib
, stdenv
, fetchurl
, libGL
, zlib
, libxkbcommon
, glib
, makeWrapper
, xorg
, dbus
, fontconfig
, freetype
, libsecret
, openssl
, makeDesktopItem
}:

let
  srcInfo = (builtins.fromJSON (builtins.readFile ./src.json)).${stdenv.hostPlatform.system}
    or (throw "ida-free is not available for the platform ${stdenv.hostPlatform.system}");
in

stdenv.mkDerivation rec {
  pname = "ida-free";
  inherit (srcInfo) version;

  src = fetchurl {
    inherit (srcInfo) url hash;
  };

  icon = fetchurl {
    urls = [
      "https://www.hex-rays.com/products/ida/news/8_1/images/icon_free.png"
      "https://web.archive.org/web/20221105181231if_/https://hex-rays.com/products/ida/news/8_1/images/icon_free.png"
    ];
    sha256 = "sha256-widkv2VGh+eOauUK/6Sz/e2auCNFAsc8n9z0fdrSnW0=";
  };

  desktopItem = makeDesktopItem {
    name = pname;
    exec = "ida64";
    icon = icon;
    comment = "Freeware version of the world's smartest and most feature-full disassembler";
    desktopName = "IDA Free";
    genericName = "Interactive Disassembler";
    categories = [ "Development" ];
  };

  ldLibraryPath = lib.makeLibraryPath ([
    stdenv.cc.cc
    zlib
    glib
    libxkbcommon
    dbus
    fontconfig
    freetype
    libGL
    libsecret
    openssl
  ] ++ (with xorg; [
    libXi
    libxcb
    libXrender
    libXau
    libX11
    libSM
    libICE
    libXext
    xcbutilwm
    xcbutilimage
    xcbutilkeysyms
    xcbutilrenderutil
  ]));

  nativeBuildInputs = [ makeWrapper ];

  dontUnpack = true;
  dontStrip = true;
  dontPatchELF = true;

  installPhase = ''
    runHook preInstall

    IDADIR=$out/opt/${pname}

    install -d $IDADIR $out/bin $out/share/applications

    # Invoke the installer with the dynamic loader directly, avoiding the need
    # to copy it to fix permissions and patch the executable.
    $(cat $NIX_CC/nix-support/dynamic-linker) ${src} \
      --mode unattended --prefix $IDADIR --installpassword ""

    rm $IDADIR/[Uu]ninstall*

    for bb in ida64 assistant; do
      patchelf --set-interpreter $(cat $NIX_CC/nix-support/dynamic-linker) $IDADIR/$bb
    done

    wrapProgram \
      $IDADIR/ida64 \
      --set LD_LIBRARY_PATH "$IDADIR:${ldLibraryPath}" \
      --set QT_PLUGIN_PATH "$IDADIR/plugins/platforms"

    ln -s $IDADIR/ida64 $out/bin/ida64

    cp $desktopItem/share/applications/* $out/share/applications

    runHook postInstall
  '';

  meta = {
    description = "Freeware version of the world's smartest and most feature-full disassembler";
    homepage = "https://hex-rays.com/ida-free/";
    sourceProvenance = with lib.sourceTypes; [ binaryNativeCode ];
    license = lib.licenses.unfree;
    maintainers = with lib.maintainers; [ athre0z ];
    platforms = [ "x86_64-linux" ];
    mainProgram = "ida64";
  };
}
```

Reply:
I would like an update script. This one is a bit convoluted though. Also who wrote it copyrightwise?

Reply:
i wrote it, its public domain. yepp, could be shorter ...

i updated versionDate so now the version is 8.2.230124